django/db/models/deletion.py

```diff
@@ -51,6 +52,23 @@ def DO_NOTHING(collector, field, sub_objs, using):
     pass
 
 
+def get_candidate_relations_to_delete(opts):
+    # Collect models that contain candidate relations to delete. This may include
+    # relations coming from proxy models.
+    candidate_models = {opts}
+    candidate_models = candidate_models.union(opts.concrete_model._meta.proxied_children)
+    # For each model, get all candidate fields.
+    candidate_model_fields = chain.from_iterable(
+        opts.get_fields(include_hidden=True) for opts in candidate_models
+    )
+    # Keep related objects not coming from ManyToManyFields from all candidate
+    # model fields.
+    return (
+        f for f in candidate_model_fields
+        if f.auto_created and not f.concrete and not f.many_to_many
```

Comment:
@loic - `GenericRelation` it's certainly a many-to-one relation, but it differs from a "native" ManyToOneField on the basis that `GenericRelation` is a lookup aid for `GenericForeignKey`, whereas a "native" ManyToOneField would be a field that actually adds a column to the related model. 

I'm not saying that a common base class can't/shouldn't be used - but there would need to be some way to tell the two cases apart. 

That said, It's difficult to account for cases where there isn't any actual code, which leads me to suspect the "explicit, but probably wrong for cases we haven't seen yet" (i.e., @collinanderson's suggestion) is probably the best approach.